staging/src/k8s.io/apiserver/pkg/audit/context.go

```diff
@@ -65,8 +60,7 @@ type annotation struct {
 // prefer AddAuditAnnotation over LogAnnotation to avoid dropping annotations.
 func AddAuditAnnotation(ctx context.Context, key, value string) {
 	ac := AuditContextFrom(ctx)
-	if ac == nil {
-		// auditing is not enabled
+	if !ac.Enabled() {
```

Comment:
I don't think `ac == nil` can be replaced with `ac.Enabled`. If the handler chain is setup properly we should never see `ac` to be `nil`, irrespective of whether audit logging is disabled. 

`// auditing is not enabled` is probably misleading, I apologize for that.  it's more like, `we should never be here if the handler chain is setup properly`

`WithAudit` filter is where we decide whether audit logging is enabled/disabled, but some filters that precede `WithAudit` filter need to add some annotations, esp. during authentication, for example:

https://github.com/kubernetes/kubernetes/blob/4db6bde859d912e11ab081a06e8ea17b2e044f5d/staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go#L136

These annotations are added before we decide whether audit logging is enabled. `ac.Enabled` is going to return `false` before that decision is made, and we are going to drop these annotations which is not expected behavior I believe.

Reply:
Ah, yeah, this ties in to the above discussion about how `Enabled()` should behave before the policy is evaluated. The approach I took is to treat it as enabled before the policy is evaluated, for exactly the reason you describe here. The other approach is to default to disabled, and use a `nil`-check here.

The advantage to my approach is once we have evaluated the policy, and explicitly decided to disabled auditing (level None), then we can skip some work here. In practice, the difference is probably negligible.

WDYT?

Reply:
I think that test case is already covered by `TestEnabled`